pkg/testing/fixture.go

```diff
@@ -714,24 +714,99 @@ func (e *ExecErr) Unwrap() error {
+	return func(opt *statusOpts) {
+		opt.noRetry = true
+	}
+}
+
+// WithRetryTimeout adjusts the retry timeout from the default value of one minute.
+func WithRetryTimeout(duration time.Duration) func(opt *statusOpts) {
+	return func(opt *statusOpts) {
+		opt.retryTimeout = duration
+	}
+}
+
+// WithRetryInterval adjusts the retry interval from the default value of one second.
+func WithRetryInterval(duration time.Duration) func(opt *statusOpts) {
+	return func(opt *statusOpts) {
+		opt.retryInterval = duration
 	}
+}
 
-	return status, nil
+// WithCmdOptions adjusts the options of the command when status is called.
+func WithCmdOptions(cmdOptions ...process.CmdOption) func(opt *statusOpts) {
+	return func(opt *statusOpts) {
+		opt.cmdOptions = append(opt.cmdOptions, cmdOptions...)
+	}
+}
+
+// ExecStatus executes `elastic-agent status --output=json`.
+//
+// Returns the parsed output and the error from the execution. Keep in mind the agent exits with status 1 if it's
+// unhealthy, but it still outputs the status successfully. This call does require that the Elastic Agent is running
+// and communication over the control protocol is working.
+//
+// By default, retry logic is applied. Use WithNoRetry to disable this behavior. WithRetryTimeout and WithRetryInterval
+// can be used to adjust the retry logic timing. The default retry timeout is one minute and the default retry
+// interval is one second.
+//
+// An empty AgentStatusOutput and non nil error means the output could not be parsed. As long as we get some output,
+// we don't return any error. It should work with any 8.6+ agent
+func (f *Fixture) ExecStatus(ctx context.Context, opts ...statusOpt) (AgentStatusOutput, error) {
+	var opt statusOpts
+	opt.retryTimeout = 1 * time.Minute
+	opt.retryInterval = 1 * time.Second
+	for _, o := range opts {
+		o(&opt)
+	}
+
+	var cancel context.CancelFunc
+	if opt.noRetry || opt.retryTimeout == 0 {
+		ctx, cancel = context.WithCancel(ctx)
+	} else {
+		ctx, cancel = context.WithTimeout(ctx, opt.retryTimeout)
+	}
+	defer cancel()
+
+	var lastErr error
+	for {
+		if ctx.Err() != nil {
+			if errors.Is(ctx.Err(), context.DeadlineExceeded) && lastErr != nil {
+				// return the last observed error
+				return AgentStatusOutput{}, fmt.Errorf("agent status returned an error: %w", lastErr)
+			}
+			return AgentStatusOutput{}, fmt.Errorf("agent status failed: %w", ctx.Err())
+		}
+		out, err := f.Exec(ctx, []string{"status", "--output", "json"}, opt.cmdOptions...)
+		status := AgentStatusOutput{}
+		if uerr := json.Unmarshal(out, &status); uerr != nil {
+			// unmarshal error means that json was not outputted due to a communication error
+			lastErr = fmt.Errorf("could not unmarshal agent status output: %w:\n%s", errors.Join(uerr, err), out)
+		} else if status.IsZero() {
+			// still not correct try again for a successful status
+			lastErr = fmt.Errorf("agent status output is empty: %w", err)
+		} else {
+			return status, nil
```

Comment:
in my code proposal (sorry for the missing `}` in the UI that was my bad) which is now here, I do make the assumption that we don't want to return the `err` from the `Exec` if we manage to unmarshal the status. I still believe that this is what we are after, but I am gonna leave this comment to make sure that you agree with this @blakerouse . Feel free to just resolve if you do.

Reply:
Don't apologize for the missing `}`! I appreciate the code you provided, it was cleaner and less duplicated. You are correct that if we can successfully read the status and its not zero that we want to return it. That is on purpose because the `status` subcommand can return non-zero if the status is not healthy. We do not want that to be the case here.

Looks like the CI is passing and this is working good now.